### ISSUE TITLE
Question about futures that do not use the waker

### ISSUE THREAD

**Issue by guswynn** [stjepang/smol#57]
One thing that has consistently confused me about futures is this quote from the docs:
https://doc.rust-lang.org/std/future/trait.Future.html

> Once a future has finished, clients should not poll it again.
>
> When a future is not ready yet, poll returns Poll::Pending and stores a clone of the Waker copied from the current Context. This Waker is then woken once the future can make progress. For example, a future waiting for a socket to become readable would call .clone() on the Waker and store it. When a signal arrives elsewhere indicating that the socket is readable, Waker::wake is called and the socket future's task is awoken. Once a task has been woken up, it should attempt to poll the future again, which may or may not produce a final value.
>
> Note that on multiple calls to poll, only the Waker from the Context passed to the most recent call should be scheduled to receive a wakeup.

To me, its clear why a PROPER future implementation should use the waker to unblock the executor (some combinators like Map don't do this, but they usually contain futures that do use the waker), but its also clear that there is nothing really stopping someone from writing a naive future that just does cpu bound work.

To test this I made https://github.com/guswynn/smol-test-block, which implements a future that yields some amount of times, then yields `Ready()` after that many times, and NEVER interacts with the `_cx`, to see if the the executor would be blocked.

```
> cargo run 10
   Compiling smol_test v0.1.0 (/Users/azw/ruststuff/smol_test)
    Finished dev [unoptimized + debuginfo] target(s) in 0.87s
     Running `target/debug/smol_test 10`
Current countdown: 9
Current countdown: 8
<hang>
```

I originally thought I'd have to countdown 200 times (as I believe that thats the number of poll's the 2 executors run on a loop, but in fact, it seemed to get stuck after only 2 poll's

It feels bad that what seems like an innocent future implementation, can hang an executor. Is something wrong with my code here? Am I totally off base thinking that futures that don't use the waker should work? To me its weird in rust for a constraint like this to not be statically enforced.

For posterity, the tokio-test branch of that repo tries the same thing, but tokio seemed to have similar behavior:
```
> cargo run 10000
   Compiling smol_test v0.1.0 (/Users/azw/ruststuff/smol_test)
    Finished dev [unoptimized + debuginfo] target(s) in 0.78s
     Running `target/debug/smol_test 10000`
Current countdown: 9999
<hang>
```

**Comment by Koxiaet** [comment]
It's impossible to _force_ Future implementations to call the waker; you can't tell the difference between a computation that takes a long time to complete and one that will never complete, like yours. Not using the waker is indistinguishable from using the waker 1000 years in the future. That's why the executor hangs.

**Comment by guswynn** [comment]
@Koxiaet 
> you can't tell the difference between a computation that takes a long time to complete and one that will never complete, like yours. Not using the waker is indistinguishable from using the waker 1000 years in the future.

This makes a lot of sense! Thanks for the response!

I guess though, I still think there may be value in having some (relatively long) timeout, after which the executor un-hangs. This would not affect perf as the timeout would be longer than normal wakeups in normal workflows, but would unblock pathological case like this.

One argument against this is that offering this papers over the pathological problems that shouldn't be written in the first place

**Comment by Ekleog** [comment]
The `Future` contract does specify that the `Waker` must be called, so your code is invalid indeed, as per the paragraph you quoted.

If you need to yield, as you say, you should use something like https://github.com/stjepang/smol/issues/43, and do something like that:
```rust
async fn foo() {
    bar();
    yield_now().await;
    baz();
}
```

Hand-written futures are usually something you don't need to do, nor want as it's quite hard to call the waker as it has to be -- either way, if you really have to write a future, the thing you must remember when writing `Future`s is that if you return `Poll::Pending`, then you *must* call the `Waker` at some point.

As for adding implicit timeouts… adding timeouts on each pending would probably slow down quite a lot the code, because it needs to record and keep track of all the implicit timeouts. And given the `Future` contract requires futures to call the waker, there's little point in doing so :)

**Comment by guswynn** [comment]
@Ekleog thanks for linking that issue! ill look into yield_now()!

**Comment by stjepang** [comment]
cc #43

**Comment by stjepang** [comment]
Closing in favor of #43
